File: wgun828.js
// William Guntur wgun828 Compsci 335 Assignment 1
// Node JS solution only using linq-es2015 (external) and fs (internal).
// Tested on lab computers on 30/07/18.

var fs = require('fs')
var Enumerable = require('linq-es2015')

function Read(){
    let fname = process.argv[2]; 
    let text = fs.readFileSync(fname, 'utf8').toUpperCase()
    let k = 3 // By default, the number of lines 'k' is 3.
    if (process.argv[3] !== undefined){
        k = process.argv[3] // If the argument 'k' is provided, 'k' will be changed to what is given.
    }
    let words = text.replace(/[^a-zA-Z ]/g, " ").trim()
    let words2 = words.replace(/\s+/g, ' ').trim().split(' ')
    let frequency = Frequencies(words2, k)
    frequency.forEach(s => console.log(s[0].toString(), s[1]))
} 

function Frequencies(words, lines){
    let frs = Enumerable.asEnumerable(words)
        .GroupBy (s => s.toUpperCase()) 
        .Select (g => [g.length, g.key]) 
        .OrderByDescending (kc => kc[0])
        .ThenBy(kc => kc[1])
        .Take(lines) // Limit the number of lines output to 'k'.
        .ToArray ()
      return frs
} 

function Main(){
    try{
        if (process.argv[2] === undefined) {
            throw new Error("Please enter a filename.") // If no filename is given.
        } else if (fs.existsSync(process.argv[2]) == false){
            throw new Error("File does not exist.") // If a filename is specified but does not exist in current directory.
        } if (process.argv[3] !== undefined){
            if (isNaN(process.argv[3])){
                throw new Error("Please enter a valid number for the number of lines to print.")  // If a non-numeric or negative input is given for 'k'.
            } else if (process.argv[3] < 0) {
                throw new Error("Please enter a non-negative number.")
            }
        }
        Read()
    } catch(ex) {
        console.log('*** Error: ' + ex.message)
    }
}
Main()

File: wgun828.cs
// William Guntur wgun828 Compsci 335 Assignment 1
// C# solution using LINQ and lambda expressions. No for, foreach, while statements used aside from Array.ForEach using lambda expression as allowed by the lecturers.
// Tested on lab computers on 30/07/18.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Text.RegularExpressions;

public class wgun828{
  public static void Main(string[] args){
    try{
      if (args == null || args.Length == 0){
        throw new Exception("Argument cannot be empty. Please enter a filename."); // Throws an error if no arguments are given.
      }
      if (!File.Exists(args[0])){
        throw new Exception("This file doesnt exist in the current directory."); // Throws an error if the filename doesnt exist in the current directory.
      }
      if (args.Length > 1){
        if (Regex.IsMatch(args[1], @"^\d+$") != true){ 
          throw new Exception("Please only enter a valid non-negative number."); // Throws an error if the second argument given is not numeric.
        }
      }
      int lines = 3; // By default, 'k' is set to 3.
      if (args.Length == 2){
        lines = int.Parse(args[1]); // Lines is 'k' as given in the assignment spec.
      }
      var fname = args[0];
      var text = File.ReadAllText(fname);
      Regex rgx = new Regex(@"[^a-zA-Z]+");
      var words = rgx.Replace(text, " ").Trim().Split(); // Replaces all non-alphabet characters with a single space.
      var wordFrequency = Frequencies (words, lines); // Call the function Frequencies, with lines as argument 'k'
      Array.ForEach(wordFrequency.ToArray(), x => Console.WriteLine("{0} {1}", x.Item1, x.Item2) ); // Using an Array.ForEach with lambda expression, allowed by lecturers on Piazza.
    } catch (Exception ex) {
           Console.WriteLine($"*** Error: {ex.Message}"); 
      Environment.ExitCode = 1; 
    } 
    IEnumerable<(int, string)> Frequencies (string[] words, int lines){
      var frs = words
        .GroupBy (s => s.ToUpper()) 
        .Select (g => (g.Count(), g.Key)) 
        .OrderByDescending (kc => kc.Item1)
        .ThenBy(kc => kc.Item2) // Second ordering.
        .Take(lines) // Implementing the second argument 'k'.
        ;
      return frs;
    }
  }
}